src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Raven.Server.Documents.Queries.AST;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using static Raven.Server.Documents.Queries.GraphQueryRunner;
+
+namespace Raven.Server.Documents.Queries.Graph
+{
+    public class EdgeQueryStep : IQueryStep
+    {
+
+        public EdgeQueryStep(IQueryStep left, IQueryStep right, WithEdgesExpression edgesExpression, MatchPath edgePath, BlittableJsonReaderObject queryParameters)
+        {
+            _left = left;
+            _right = right;
+            _edgePath = edgePath;
+            _queryParameters = queryParameters;
+            _edgesExpression = edgesExpression;
+        }
+
+        public async ValueTask<IEnumerable<Match>> Execute(Dictionary<IQueryStep, IEnumerable<Match>> matches)
+        {
+            var leftResult = await _left.Execute(matches);
+            var rightResult = await _right.Execute(matches);
```

Comment:
Here you need to check if they are completed or wait.

Reply:
just to be clear i got you right, you want to remove the async method and replace it with.Wait()?

Reply:
the task themselves will never be completed at this point since the execution is reccursive in a post order manner.